tap_mailchimp/streams.py

```diff
@@ -0,0 +1,601 @@
+        offset = 0
+        has_more = True
+        params = {
+            'count': page_size,
+            **self.params
+        }
+        while has_more:
+            params['offset'] = offset
+
+            # Add param for querying records after a particular date based on the 'bookmark_query_field'
+            if self.bookmark_query_field:
+                params[self.bookmark_query_field] = last_datetime
+
+            LOGGER.info(
+                '%s - Syncing - %scount: %s, offset: %s',
+                self.stream_name,
+                'since: {}, '.format(last_datetime) if self.bookmark_query_field else '',
+                page_size,
+                offset
+            )
+
+            formatted_selected_fields = self.format_selected_fields()
+
+            params['fields'] = formatted_selected_fields
+            data = self.client.get(
+                self.path,
+                params=params,
+                endpoint=self.stream_name)
+
+            raw_records = data.get(self.data_key)
+
+            if len(raw_records) < page_size:
+                has_more = False
+
+            # Loop over every record and sync child stream
+            for record in raw_records:
+                # Remove '_links' as it contains API schema docs
+                del record['_links']
+                # Store 'ids' for reports streams
+                ids.append(record.get('id'))
+                if self.child:
+                    for child in self.child:
+                        # Skip reports child stream sync as it syncs for the list of parent ids
+                        if child in self.report_streams:
+                            continue
+                        # If the child stream is selected or the grandchild is selected then sync the child stream
+                        if child in self.selected_stream_names or child in self.child_streams_to_sync:
+                            self.sync_substream(child, record.get('id'))
+
+            # If the stream is selected then write records
+            if self.to_write_records:
+                max_bookmark_field = self.process_records(raw_records, max_bookmark_field, sync_start_date)
+
+            # Sync 'reports' stream based on 'ids' collected for every record
+            for report_stream in self.report_streams:
+                # Skip the sync if the stream is not selected
+                if report_stream not in self.selected_stream_names:
+                    continue
+                reports_stream_obj = STREAMS.get(report_stream)(self.state, self.client, self.config, \
+                    self.catalog, self.selected_stream_names, self.child_streams_to_sync)
+                reports_stream_obj.sync_report_activities(ids)
+
+            # Write bookmark for Incremental streams
+            if self.bookmark_query_field or self.stream_name in ["unsubscribes"]:
+                self.write_bookmark(self.bookmark_path, max_bookmark_field)
```

Comment:
The base stream class should be a generic implementation of the stream and should not have any dependancy over the classes inheriting it, hence a condition specific to the child class `unsubscribes` should not be a part of this implementation
this should be done by overriding this method in the child class.